sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEOptimizer.scala

```diff
@@ -27,7 +28,9 @@ import org.apache.spark.util.Utils
  */
 class AQEOptimizer(conf: SQLConf) extends RuleExecutor[LogicalPlan] {
   private val defaultBatches = Seq(
-    Batch("Eliminate Unnecessary Join", Once, EliminateUnnecessaryJoin),
+    Batch("Propagate Empty Relations", Once,
+      AQEPropagateEmptyRelation,
+      UpdateAttributeNullability),
```

Comment:
shall we also put `ConvertToLocalRelation` in this batch?

Reply:
yeah, I noticed it. We can put it so that we can propagate empty through `project/filter`. like such case:
```
Aggregate
  Project
    Join
      Shuffle
```
But it need to isolate normal and AQE due to `transformWithPruning`.

Otherhand I feel that it's similar if we just let `AQEPropagateEmptyRelation` support propagate `project/filter`. and the later is simpler.

Reply:
It's a bit different:
```
Project
  Shuffle Stage
```
For the above case, we don't want to optimize it as the benefit is too small

```
Project
  Sort
    Shuffle Stage
```
For the above case, we will optimize Sort -> Shuffle Stage to empty relation first. Then it makes sense to optimize further and optimize out project, as the shuffle stage is already gone.

So adding `ConvertToLocalRelation` looks the best solution here.

Reply:
ah I see, will do this.